File: android/app/src/main/kotlin/org/gaindrive/android/data/model/LibraryMode.kt
package org.gaindrive.android.data.model

/**
 * Which slice of the library the top-level list is showing.
 *
 * A value class over a plain string rather than an enum, deliberately. The
 * slices on offer come from the server, so an enum would have to be edited every
 * time a server grows a new kind of root and would have no sensible branch for
 * one it had never heard of.
 *
 * A list must never mix slices, which is why this is a single value rather than
 * a set: the server filters on it and the resulting list is single-slice by
 * construction.
 *
 * The string names one of two different things, distinguished by a prefix:
 *
 * * a **content type** — `artists`, `categories` — which is a gaindrive
 *   extension naming a *kind* of root, and may span several of them
 * * a **music folder**, written `folder:<name>`, which is one root of a server
 *   that has no concept of kinds
 *
 * Keeping both in one string is what lets `SettingsStore.libraryMode` and the
 * mirror's `contentType` column carry either without a migration, and what makes
 * a value written by an older build still parse. A folder is keyed on its
 * **name** rather than its id because ids are per-server and the chip row unions
 * several servers — the same reason `Merge.kt` matches everything else by name.
 *
 * Personal uploads are not a slice here, although a URL shared with the app now
 * lands in them — see `ui/fetch/`. Browsing them is the part still missing, and
 * a chip is not the whole of it: the listing needs `personal=true` on every
 * hierarchy query and an admin-only promote action, none of which the merge and
 * mirror layers below have a notion of yet.
 */
@JvmInline
value class LibraryMode(val id: String) {

	/** The folder this names, or null when it names a content type. */
	val folderName: String?
		get() = if (id.startsWith(FOLDER)) id.removePrefix(FOLDER) else null

	/**
	 * Title-cased for display. A folder shows its own name, which the server's
	 * owner chose and is already how they think of it. A content type falls back
	 * to capitalising the raw id, so a kind this app has never seen still shows
	 * something sensible rather than nothing.
	 */
	val label: String
		get() = folderName ?: (LABELS[id] ?: id.replaceFirstChar { it.uppercase() })

	companion object {
		/**
		 * What a server that names no kind of root is taken to hold — see
		 * `data/browse/LibraryRoots.kt`.
		 */
		val ARTISTS = LibraryMode("artists")

		/**
		 * A folder named "artists" is `folder:artists`, so it cannot be mistaken
		 * for the content type of the same name.
		 */
		fun folder(name: String) = LibraryMode("$FOLDER$name")

		private const val FOLDER = "folder:"

		private val LABELS = mapOf(
			"artists" to "Artists",
			"categories" to "Categories",
		)
	}
}

File: ios/GainDrive/Data/Model/LibraryMode.swift
//	GainDrive for iOS
//	Copyright (C) 2026 Kasper Peeters
//
//	This file is part of GainDrive, distributed under the GNU General
//	Public License version 3 or later, with the Application Distribution
//	Exception in ios/LICENSE. See LICENSE at the repository root for the
//	full text of the GPL.

import Foundation

/// Which slice of the library the top-level list is showing.
///
/// **A wrapper over a plain string rather than an enum**, deliberately, and for
/// the reason `data/model/LibraryMode.kt` gives: the slices on offer come from
/// the server, so an enum would have to be edited every time a server grew a
/// new kind of root and would have no sensible case for one it had never heard
/// of.
///
/// A listing must never mix slices, which is why this is a single value rather
/// than a set: the server filters on it, so the result is single-slice by
/// construction.
///
/// The string names one of two things today:
///
/// * a **content type** — `artists`, `categories` — a gaindrive extension
///   naming a *kind* of root, which may span several of them
/// * **`uploads`**, the account's own upload area
///
/// Android has a third spelling, `folder:<name>`, for a server browsing by
/// folder. That arrives with folder browsing itself; the string is the reason
/// it will need no migration when it does, since `SettingsStore` stores this
/// value as it stands and an unrecognised one already falls back.
struct LibraryMode: Hashable, Sendable, Identifiable, Codable {
	let id: String

	init(_ id: String) {
		self.id = id
	}

	/// What a server naming no kind of root is taken to hold.
	static let artists = LibraryMode("artists")

	/// Sections rather than performers — Film, Series. A content type like any
	/// other; named here because the artist header keys on it: a section has no
	/// portrait and no biography, and the server refuses to look one up
	/// (`is_category_folder()`).
	static let categories = LibraryMode("categories")

	/// The account's own upload area.
	///
	/// **Looks like a content type and is not one.** The server deliberately
	/// keeps its uploads root out of `getMusicFolders`, so no amount of
	/// inspecting the roots will ever produce this chip — it is offered because
	/// the *account* may upload. It reaches the wire as `personal=true` rather
	/// than as a `contentType`, which is the whole of what `LibraryRoots` has
	/// to special-case.
	static let uploads = LibraryMode("uploads")

	/// Title-cased for display, falling back to capitalising the raw id so a
	/// kind this app has never seen still shows something rather than nothing.
	var label: String {
		if let known = Self.labels[id] { return known }
		guard let first = id.first else { return id }
		return first.uppercased() + id.dropFirst()
	}

	private static let labels = [
		"artists": "Artists",
		"categories": "Categories",
		"uploads": "Uploads",
	]
}
